File: eww/ws.py
#!/usr/bin/pypy3

import subprocess
import json

def get_dwm_workspace_names():
    dwm_status = subprocess.run(
        ["xprop", "-root", "_NET_DESKTOP_NAMES"],
        capture_output=True,
        text=True,
    ).stdout

    if "_NET_DESKTOP_NAMES" in dwm_status:
        workspace_string = dwm_status.split('=')[-1].strip()
        workspace_string = workspace_string.strip('"')
        return [ {
            "idx":  idx+1,
            "name": workspace.strip(),
            "class": "inactive"
        } for idx, workspace in enumerate(workspace_string.split('", "'))]
    else:
        return []

def active_ws():
    wmctrl_output = subprocess.run(
        ["wmctrl", "-d"],
        capture_output=True,
        text=True,
    ).stdout

    lines = wmctrl_output.splitlines()
    for line in lines:
        if '*' in line:  # This line represents the active workspace
            return int(line.split()[0])  # The workspace number is the first field

    return None  # Return None if no active workspace was found



def main():
    workspaces = get_dwm_workspace_names()
    active_workspace = active_ws()

    if active_workspace is not None:
        for workspace in workspaces:
            if workspace["idx"] == active_workspace + 1:  # Indices in wmctrl start at 0
                workspace["class"] = "active"
                break

    workspaces = json.dumps(workspaces)
    print(workspaces)


if __name__ == "__main__":
    main()

File: eww/go/ws.go
package main

import (
	"encoding/json"
	"fmt"
	"os/exec"
	"strconv"
	"strings"
)

type Workspace struct {
	Idx   int    `json:"idx"`
	Name  string `json:"name"`
	Class string `json:"class"`
}

func getDwmWorkspaceNames() ([]Workspace, error) {
	cmd := exec.Command("xprop", "-root", "_NET_DESKTOP_NAMES")
	output, err := cmd.Output()
	if err != nil {
		return nil, err
	}

	if strings.Contains(string(output), "_NET_DESKTOP_NAMES") {
		workspaces := make([]Workspace, 0)
		workspaceString := strings.Split(string(output), "=")[1]
		workspaceString = strings.Trim(workspaceString, "\"\n ")
		for idx, workspace := range strings.Split(workspaceString, "\", \"") {
			workspaces = append(workspaces, Workspace{Idx: idx + 1, Name: workspace, Class: "inactive"})
		}
		return workspaces, nil
	}

	return nil, nil
}

func activeWs() (int, error) {
	cmd := exec.Command("wmctrl", "-d")
	output, err := cmd.Output()
	if err != nil {
		return 0, err
	}

	lines := strings.Split(string(output), "\n")
	for _, line := range lines {
		if strings.Contains(line, "*") {
			fields := strings.Fields(line)
			activeWorkspace, err := strconv.Atoi(fields[0])
			if err != nil {
				return 0, err
			}
			return activeWorkspace, nil
		}
	}
	return 0, nil
}

func main() {
	workspaces, err := getDwmWorkspaceNames()
	if err != nil {
		fmt.Println(err)
		return
	}

	activeWorkspace, err := activeWs()
	if err != nil {
		fmt.Println(err)
		return
	}

	for i := range workspaces {
		if workspaces[i].Idx == activeWorkspace+1 {
			workspaces[i].Class = "active"
			break
		}
	}

	jsonData, err := json.Marshal(workspaces)
	if err != nil {
		fmt.Println(err)
		return
	}
	fmt.Println(string(jsonData))
}
